File: 328/odd-even-linked-list.go
type ListNode struct {
	Val  int
	Next *ListNode
}

func oddEvenList(head *ListNode) *ListNode {
	if head == nil {
		return head
	}

	// odd group's end
	oe := head
	// even group's begin
	eb := head.Next
	// even group's end
	ee := eb

	for ee != nil && ee.Next != nil {
		oe.Next = ee.Next
		oe = oe.Next
		ee.Next = oe.Next
		ee = ee.Next
		oe.Next = eb
	}

	return head
}

File: 328/odd-even-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var oddEvenList = function (head) {
    if (head == null) {
        return head;
    }

    let oe = head;
    let eb = head.next;
    let ee = eb;

    while (ee != null && ee.next != null) {
        oe.next = ee.next;
        oe = oe.next;
        ee.next = oe.next;
        ee = ee.next;
        oe.next = eb;
    }

    return head;
};